src/Databases/DatabaseOnDisk.cpp

```diff
@@ -673,6 +679,69 @@ ASTPtr DatabaseOnDisk::getCreateQueryFromMetadata(const String & database_metada
     return ast;
 }
 
+ASTPtr DatabaseOnDisk::getCreateQueryFromStorage(const String & table_name, const StoragePtr & storage, bool throw_on_error) const
+{
+    auto metadata_ptr = storage->getInMemoryMetadataPtr();
+    if (metadata_ptr == nullptr)
+    {
+        if (throw_on_error)
+            throw Exception(ErrorCodes::CANNOT_GET_CREATE_TABLE_QUERY, "Cannot get metadata of {}.{}", backQuote(getDatabaseName()), backQuote(table_name));
+        else
+            return nullptr;
+    }
+
+    auto create_table_query = std::make_shared<ASTCreateQuery>();
+    create_table_query->setTable(table_name);
+    create_table_query->setDatabase(getDatabaseName());
+    create_table_query->attach = false;
+    create_table_query->set(create_table_query->comment, std::make_shared<ASTLiteral>("SYSTEM TABLE is built on the fly."));
+
+    /// setup create table query storage info.
+    {
+        auto ast_engine = std::make_shared<ASTFunction>();
+        ast_engine->name = storage->getName();
+        auto ast_storage = std::make_shared<ASTStorage>();
+        ast_storage->set(ast_storage->engine, ast_engine);
+        create_table_query->set(create_table_query->storage, ast_storage);
+    }
+    /// setup create table query columns info.
```

Comment:
Btw, we already have similar code [here](https://github.com/ClickHouse/ClickHouse/blob/7b5954b736465aa059803139a3870aeab3d2d327/src/Databases/MySQL/DatabaseMySQL.cpp#L116) and [here](https://github.com/ClickHouse/ClickHouse/blob/d76976e272268f76bb469d611aca66acd9dd1cbf/src/Databases/SQLite/DatabaseSQLite.cpp#L173). Is it possible to extract common code to some function or static method?

Reply:
Ok, I will have a try.